src/Core/Framework/Gateway/Context/Command/ContextGatewayCommandCollection.php

```diff
@@ -0,0 +1,34 @@
+<?php declare(strict_types=1);
+
+namespace Shopware\Core\Framework\Gateway\Context\Command;
+
+use Shopware\Core\Framework\Log\Package;
+use Shopware\Core\Framework\Struct\Collection;
+
+/**
+ * @extends Collection<AbstractContextGatewayCommand>
+ */
+#[Package('framework')]
+class ContextGatewayCommandCollection extends Collection
+{
+    public function getSingleTokenCommand(): LoginCustomerCommand|RegisterCustomerCommand|null
+    {
+        /** @var LoginCustomerCommand|RegisterCustomerCommand|null $command */
+        $command = $this->getTokenCommands()->first();
+
+        return $command;
+    }
+
+    public function getTokenCommands(): self
```

Comment:
To type this correctly, you can this:
```suggestion
/**
 * @template T of AbstractContextGatewayCommand = AbstractContextGatewayCommand
 *
 * @extends Collection<T>
 */
#[Package('framework')]
class ContextGatewayCommandCollection extends Collection
{
    public function getSingleTokenCommand(): LoginCustomerCommand|RegisterCustomerCommand|null
    {
        return $this->getTokenCommands()->first();
    }

    /**
     * @return static<LoginCustomerCommand|RegisterCustomerCommand>
     */
    public function getTokenCommands(): static
```

Reply:
this is still valid, `getTokenCommands` should still return `self<LoginCustomerCommand|RegisterCustomerCommand>` and the type assertion within `getSingleTokenCommand` is no longer needed

Reply:
It does not work with `static`, but I pointed it to `self`, which phpstan accepts.